flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/allocator/DefaultSlotAssignerTest.java

```diff
@@ -0,0 +1,121 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.apache.flink.runtime.scheduler.adaptive.allocator.DefaultSlotAssigner.APPLICATION_MODE_EXECUTION_TARGET;
+import static org.apache.flink.runtime.scheduler.adaptive.allocator.JobInformation.VertexInformation;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for {@link DefaultSlotAssigner}. */
+@ExtendWith(ParameterizedTestExtension.class)
+class DefaultSlotAssignerTest {
+
+    private static final TaskManagerLocation tml1 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml1 = new TestSlotInfo(tml1);
+    private static final SlotInfo slot2OfTml1 = new TestSlotInfo(tml1);
+    private static final SlotInfo slot3OfTml1 = new TestSlotInfo(tml1);
+
+    private static final TaskManagerLocation tml2 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml2 = new TestSlotInfo(tml2);
+    private static final SlotInfo slot2OfTml2 = new TestSlotInfo(tml2);
+    private static final SlotInfo slot3OfTml2 = new TestSlotInfo(tml2);
+
+    private static final TaskManagerLocation tml3 = new LocalTaskManagerLocation();
+    private static final SlotInfo slot1OfTml3 = new TestSlotInfo(tml3);
+    private static final SlotInfo slot2OfTml3 = new TestSlotInfo(tml3);
+
+    private static final List<SlotInfo> allSlots =
+            Arrays.asList(
+                    slot1OfTml1,
+                    slot2OfTml1,
+                    slot3OfTml1,
+                    slot1OfTml2,
+                    slot2OfTml2,
+                    slot3OfTml2,
+                    slot1OfTml3,
+                    slot2OfTml3);
+
+    private static final JobVertex jobVertex = new JobVertex("testingJobVertex");
+    private static final SlotSharingGroup slotSharingGroup = new SlotSharingGroup();
+
+    @Parameter int parallelism;
+
+    @Parameter(value = 1)
+    Collection<? extends SlotInfo> freeSlots;
+
+    @Parameter(value = 2)
+    List<TaskManagerLocation> minimalTaskExecutors;
+
+    @TestTemplate
+    void testAssignSlots() {
+        final SlotAssigner slotAssigner =
+                new DefaultSlotAssigner(APPLICATION_MODE_EXECUTION_TARGET, true);
+        final VertexInformation vertexInfo =
+                new TestVertexInformation(jobVertex.getID(), parallelism, slotSharingGroup);
+        final VertexParallelism vertexParallel =
+                new VertexParallelism(
+                        singletonMap(vertexInfo.getJobVertexID(), vertexInfo.getParallelism()));
+        final JobInformation jobInformation = new TestJobInformation(singletonList(vertexInfo));
+        final Set<TaskManagerLocation> keptTaskExecutors =
+                slotAssigner.assignSlots(jobInformation, freeSlots, vertexParallel, null).stream()
+                        .map(assignment -> assignment.getSlotInfo().getTaskManagerLocation())
```

Comment:
I'm not a big fan of filtering based on the `TaskManagerLocation` here: We could achieve the same by checking what slots are in the end picked for the slot assignment. Because that's what the `SlotAssigner` actually does.